lnst/RecipeCommon/Perf/Measurements/ForwardingMeasurement.py

```diff
@@ -0,0 +1,469 @@
+"""
+Module implementing the actual Forwarding Measurement.
+
+Copyright 2025 Red Hat, Inc.
+Licensed under the GNU General Public License, version 2 as
+published by the Free Software Foundation; see COPYING for details.
+"""
+
+__author__ = """
+sdobron@redhat.com (Samuel Dobron)
+"""
+
+from typing import Literal
+
+from lnst.Tests.XDPBench import XDPBench
+from lnst.RecipeCommon.Perf.Results import (
+    PerfInterval,
+    ParallelPerfResult,
+    SequentialPerfResult,
+)
+from lnst.Devices.VlanDevice import VlanDevice
+from lnst.Controller.RecipeResults import ResultType
+from lnst.Tests.PktGen import PktgenController, PktgenDevice
+
+from lnst.Tests.InterfaceStatsMonitor import InterfaceStatsMonitor
+from lnst.RecipeCommon.Perf.Measurements.BaseFlowMeasurement import (
+    Flow,
+    NetworkFlowTest,
+)
+from lnst.RecipeCommon.Perf.Measurements.MeasurementError import MeasurementError
+from lnst.RecipeCommon.Perf.Measurements.BaseFlowMeasurement import BaseFlowMeasurement
+from lnst.RecipeCommon.Perf.Measurements.Results.ForwardingMeasurementResults import (
+    ForwardingMeasurementResults,
+)
+
+from lnst.RecipeCommon.Perf.Measurements.Results.AggregatedForwardingMeasurementResults import (
+    AggregatedForwardingMeasurementResults,
+)
+from lnst.Controller.RecipeResults import MeasurementResult
+
+
+class ForwardingMeasurement(BaseFlowMeasurement):
+    """
+    This class implements the forwarding measurement.
+    It uses pktgen to generate packets originating from
+    :attr:`Flow.generator_nic` destinating :attr:`Flow.receiver_nic`
+    via :attr:`Flow.forwarder_nic`.
+
+    :attr:`Flow.receiver_nic` runs XDP-bench in drop mode.
+    Sure, we could use some nftables rule but xdp-bench is much
+    faster and requires less separate CPUs for dropping packets
+    and so, we can use more CPUs for generating traffic.
+
+    This class expects forwarding to be already configured.
+    Underlying forwarding plane could be anything like regular
+    kernel, xdp-forward, ...
+
+    This measurement runs single instance of pktgen, multiple
+    instances of xdp-bench as well as multiple instances of
+    interface stats monitor, based on destination/forwarding
+    interfaces.
+
+    Since it's not possible to measure individual flows, there
+    are just per-interface stats monitors. Results are then
+    spread over the flows passing through that interface.
+    """
+
+    def __init__(
+        self,
+        flows,
+        ratep=-1,
+        burst=1,
+        xdp_mode: Literal["skb", "native"] = "native",
+        recipe_conf=None,
+    ):
+        super().__init__(recipe_conf=recipe_conf)
+        self._flows = flows
+        self._mode = xdp_mode
+        self._ratep = ratep
+        self._burst = burst
+
+        self._client_job = None  # pktgen
+        self._server_job = None  # xdp-bench
+        self._forwarder_job = None  # interface stats monitor
```

Comment:
forwarder job == stats monitor? naming here seems a bit misleading.

does it make sense to name this as client/server/forwarder? wouldn't it be more appropriate to call it generator/dropper/forwarder_monitor?

Reply:
> forwarder job == stats monitor?

yes

I wanted to keep it consistent across measurement but why not, that's more descriptive.